icu4c/source/tools/toolutil/writesrc.cpp

```diff
@@ -395,3 +461,29 @@ usrc_writeArrayOfMostlyInvChars(FILE *f,
         fputs(postfix, f);
     }
 }
+
+U_CAPI void U_EXPORT2
+usrc_writeString(FILE *f,
+                 const UChar* ptr, int32_t length,
+                 UTargetSyntax) {
+    UnicodeString ustr(ptr, length);
```

Comment:
```suggestion
    UnicodeString ustr(false, ptr, length);  // read-only alias
```
... but actually, since you are just iterating over the string's code points, just use the U16_NEXT(ptr, i, length, codePoint) macro instead.